packages/counterfactual_nft/contracts/WithCreator.sol

```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.8.2;
+
+
+/// @title AddressSet
+/// @author Freeman Zhong
+contract WithCreator {
```

Comment:
Instead of using this to inherit from in `CounterfactualNFT` I think it makes more sense to inherit `CounterfactualNFT` on this contract. You can disable minting here by overriding the mint function and just throw an error. You can also override `mintFromL2` to call `_setCreator` and then just pass it through to `CounterfactualNFT.mintFromL2` (which you'll have to change from `external` to `public` I think to be able to do so). This way `CounterfactualNFT` is left unchanged and still implements the default NFT logic (which includes minting on L1).